src/data/W/constructions.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2015 Joseph Hua. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Hua
+-/
+import data.W.basic
+
+/-!
+# Examples of W-types
+
+We take the view of W types as inductive types.
+Given `α : Type` and `β : α → Type`, the W type determined by this data, `W_type β`, is the
+inductively with constructors from `α` and arities of each constructor `a : α` given by `β a`.
+
+This file contains `nat` and `list` as examples of W types.
+
+## Main results
+* `W_type.equiv_nat`: the construction of the naturals as a W-type is equivalent
+  to `nat`
+* `W_type.equiv_list`: the construction of lists on a type `γ` as a W-type is equivalent to
+  `list γ`
+-/
+
+universes u v
+
+namespace W_type
+
+section nat
+
+/-- The constructors for the naturals -/
+inductive nat_α : Type
+| zero : nat_α
+| succ : nat_α
+
+instance : inhabited nat_α := ⟨ nat_α.zero ⟩
+
+/-- The arity of the constructors for the naturals, `zero` takes no arguments, `succ` takes one -/
+def nat_β : nat_α → Type
+| nat_α.zero := empty
+| nat_α.succ := unit
+
+instance : inhabited (nat_β nat_α.succ) := ⟨ () ⟩
+
+/-- The isomorphism from the naturals to its corresponding `W_type` -/
+@[simp] def of_nat : ℕ → W_type nat_β
+| nat.zero := ⟨ nat_α.zero , empty.elim ⟩
+| (nat.succ n) := ⟨ nat_α.succ , λ _ , of_nat n ⟩
+
+/-- The isomorphism from the `W_type` of the naturals to the naturals -/
+@[simp] def to_nat : W_type nat_β → ℕ
+| (W_type.mk nat_α.zero f) := 0
+| (W_type.mk nat_α.succ f) := (to_nat (f ())).succ
+
+lemma left_inv_nat : function.left_inverse of_nat to_nat
+| (W_type.mk nat_α.zero f) := by { simp, tidy }
+| (W_type.mk nat_α.succ f) := by { simp, tidy }
+
+lemma right_inv_nat : function.right_inverse of_nat to_nat
+| nat.zero := rfl
+| (nat.succ n) := by rw [of_nat, to_nat, right_inv_nat n]
+
+/-- The naturals are equivalent to their associated `W_type` -/
+def equiv_nat : W_type nat_β ≃ ℕ :=
+{ to_fun := to_nat,
+  inv_fun := of_nat,
+  left_inv := left_inv_nat,
+  right_inv := right_inv_nat }
+
+open sum punit
+
+/--
+`nat_α` is equivalent to `punit ⊕ punit`.
+This is useful when considering the associated polynomial endofunctor.
+-/
+@[simps] def nat_α_equiv_punit_sum_punit : nat_α ≃ punit.{u + 1} ⊕ punit :=
+{ to_fun := λ c, match c with | nat_α.zero := inl star | nat_α.succ := inr star end,
+  inv_fun := λ b, match b with | inl x := nat_α.zero | inr x := nat_α.succ end,
+  left_inv := λ c, match c with | nat_α.zero := rfl | nat_α.succ := rfl end,
+  right_inv := λ b, match b with | inl star := rfl | inr star := rfl end }
+
+end nat
+
+section list
+
+variable (γ : Type u)
+
+/--
+The constructors for lists.
+There is "one constructor `cons x` for each `x : γ`",
+since we view `list γ` as
+```
+```
+| nil : list γ
+| cons x₀ : list γ → list γ
+| cons x₁ : list γ → list γ
+|   ⋮      γ many times
+```
+```
```

Comment:
````suggestion
```
| nil : list γ
| cons x₀ : list γ → list γ
| cons x₁ : list γ → list γ
|   ⋮      γ many times
```
````
Looks like you applied my suggestion twice